### DISCUSSION TITLE
Non-DVR application providing DVR playback

### DISCUSSION THREAD

**Discussion by naanlizard** [OvenMediaLabs/OvenMediaEngine#1208] [Help]
As discussed elsewhere, when in the background and muted, chrome will pause a stream. This is bad for viewers, as they fall behind.

One of our attempted fixes/workarounds was to create a separate application in OME that doesn't provide DVR, so in theory shouldn't be able to fall behind.

Here's our [Server.xml](https://gist.github.com/naanlizard/876087d22c19971d67aefb8fe3466a41)

To reproduce the thing I'm talking about, start OME and forward the same stream to both applications. Play the stream from the nodvr application via llhls, and then put the stream in a background tab while muted. Come back in a few minutes and see that the stream is playing content from when you tabbed out, making requests to the server like everything is normal, and the server giving ten minute old content with no trouble.

I'm unsure if this is intended behavior or a bug, can you please assist @getroot? (Apologies, I do not know if you all get notifications for new discussions)

**Comment by bchah** [comment]
I use a client-side workaround for a different issue that might be useful here. Set up a "visibilitychange" listener in JavaScript to detect when the user tabs away or backgrounds the window - when they change back, reload the live source in the player or use seek API if you are using OvenPlayer to skip forward to now. This solved a headache where playback in Safari goes out of sync if you tab away then back.

**Comment by getroot** [comment]
When hls.js is playing HLS, even if the user clicks pause, the player continues to download new segments and buffers them. (OvenMediaEngine does not give the player the old segment. It has already been deleted from memory.) I think this is natural. I think many users expect to continue watching the live broadcast when they pause and then resume. So I think that hls.js is doing this by default. If you want the user to always see the most up-to-date stream, it would be better to receive an event and seek to the latest point, as bcah suggested. Alternatively, it would be nice to see if there is such a thing among the options of hls.js. Let's take a look at the options of hls.js. @SangwonOh

**Reply by naanlizard** [reply]
At least for our users, the flow is something like

User loads stream(s) on a page, often muted. They watch some and then tab out to do something else, then come back to check in every so often. So when the stream falls behind they are confused, if they didn't choose to pause intentionally - the auto pause in the background is confusing.

Regardless, I think you are incorrect about OME not serving the chunks - I watched the network requests when I came back, and they seemed to be older and happening while I watched, requesting old chunks. Can you please double check?

https://i.imgur.com/SWT8Lc1.png

The chrome team suggested a workaround in the bug ticket I made and said they won't fix, which I think is foolish but whatever. I will investigate our options and report back probably in the ovenplayer issue I've made. This discussion should focus on just the potential DVR content incorrectly being served from the one application.

**Reply by getroot** [reply]
Are you assuming that the player requested the latest segment, but the server responded with an older segment? I don't think so. The way HLS works is that the player first requests the chunklist and segments, and the server provides them. When the server receives a chunklist request with _HLS_msn=5738&_HLS_part=6, the server responds if there is a newer chunklist than that. If a player requests a segment that is not on the server, respond with a 404.

Depending on your settings, your application only holds segments of 60 to 120 seconds in memory. With the screenshots you captured, it's hard to understand that OME provides old segments. Your screenshot is requesting the latest chunklist, not segments. I don't 100% understand how hls.js works, but maybe it's because I don't need to request a segment yet?

Anyway, I'm pretty sure this should be checked on the hls.js side (even if this isn't the problem, check if there are other options), but I'll think about something else.

**Reply by naanlizard** [reply]
I think it is providing old segments because it was continuing to play very old stream data. I'll check again tomorrow more rigorously - but basically I got it to fall behind, then let it play for several minutes showing content from 10+ minutes prior. Then I hit "live streaming" to sync to live and the URL requested changed significantly

**Reply by getroot** [reply]
I think you might want to check what happens in the chrome dev tools when you pause the stream at https://hlsjs.video-dev.org/demo/.

**Reply by getroot** [reply]
The old segment the player plays is the stream buffered by the player.

**Reply by getroot** [reply]
![image](https://user-images.githubusercontent.com/17779085/235569384-ca38c45d-b29b-49b3-af7e-b50b809623bd.png)

**Reply by getroot** [reply]
Even in the paused state, hls.js continues to download and buffer the latest segments. And when it resumes, it starts playing again from the paused position.

**Reply by naanlizard** [reply]
But the requests appeared in the browser console AFTER I tab back in, and it is playing old stream content. I would expect the requests to occur while it was NOT focused, that is what is confusing me.

**Reply by getroot** [reply]
I thought the player paused when the tab went to the background and resumed when it came back to the foreground. isn't this?

**Reply by naanlizard** [reply]
Yes the player pauses in the background. When you go back to the tab, it resumes from the point when you tab away (in the past DVR content). The player continues requesting segments and playing the old content.

But I realize maybe it then is requesting the newest segments, not old segments. I'm not sure how I could test this. I think you're right though, that this is hls buffering things in the background.

**Comment by naanlizard** [comment]
I'll close the discussion, I think I was incorrect and it was just buffering that was happening.

When unfocused we just unpause the stream now, and that works though is inefficient. We'll keep working to see if we can improve that.
